src/groupeddataframe/grouping.jl

```diff
@@ -178,9 +191,135 @@ function Base.getindex(gd::GroupedDataFrame, idxs::AbstractArray)
     GroupedDataFrame(gd.parent, gd.cols, new_groups, gd.idx, new_starts, new_ends)
 end
 
+# Iterator over values of grouping columns for each group
+_groupvalues(gd::GroupedDataFrame) = (Tuple(gd.parent[gd.idx[s], gd.cols]) for s in gd.starts)
+
 Base.getindex(gd::GroupedDataFrame, idxs::Colon) =
     GroupedDataFrame(gd.parent, gd.cols, gd.groups, gd.idx, gd.starts, gd.ends)
 
+"""
+    keys(gd::GroupedDataFrame)
+
+Get the values of the grouping columns for each group. These can be used to get
+the groups using the dictionary interface (by passing the key to `getindex`
+or `get`). See [`get(gd::GroupedDataFrame, key, default)`](@ref).
+
+### Returns
+
+A vector of `NamedTuple`s with names corresponding to the grouping columns.
+
+### Examples
+
+```jldoctest
+julia> df = DataFrame(a = repeat([:foo, :bar, :baz], outer=[2]),
+                      b = repeat([2, 1], outer=[3]),
+                      c = 1:6);
+
+julia> gd = groupby(df, :a)
+GroupedDataFrame with 3 groups based on key: a
+First Group (2 rows): a = :foo
+│ Row │ a      │ b     │ c     │
+│     │ Symbol │ Int64 │ Int64 │
+├─────┼────────┼───────┼───────┤
+│ 1   │ foo    │ 2     │ 1     │
+│ 2   │ foo    │ 1     │ 4     │
+⋮
+Last Group (2 rows): a = :baz
+│ Row │ a      │ b     │ c     │
+│     │ Symbol │ Int64 │ Int64 │
+├─────┼────────┼───────┼───────┤
+│ 1   │ baz    │ 2     │ 3     │
+│ 2   │ baz    │ 1     │ 6     │
+
+julia> keys(gd)
+3-element Array{NamedTuple{(:a,),Tuple{Symbol}},1}:
+ (a = :foo,)
+ (a = :bar,)
+ (a = :baz,)
+```
+"""
+function Base.keys(gd::GroupedDataFrame)
+    colnames = Tuple(_names(gd)[gd.cols])
+    coltypes = [eltype(gd.parent[:, c]) for c in gd.cols]
```

Comment:
```suggestion
    coltypes = [eltype(gd.parent[!, c]) for c in gd.cols]
```

we do not need to make a copy here. Right?

Reply:
Is that what the `!` syntax does?

Reply:
`df[:, c]` copies the column and `df[!, c]` does not copy it (i.e. exposes column as it is stored in the `DataFrame` - in general it is unsafe, but here we just want to check its type and we do not modify it so it is OK, and it is much cheaper)